klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+        '''
+
+        analysis_rules.append(pass_patterns_rule)
+        analysis_rules.append(fail_patterns_rule)
+
+    with open(output_runset_path, "w") as runset_analysis:
+        runset_analysis.write("".join(analysis_rules))
+
+    return output_gds_path, output_runset_path
+
+
+def build_tests_dataframe(unit_test_cases_dir, target_table):
+    """
+    This function is used for getting all test cases available in a formated dataframe before running.
+
+    Parameters
+    ----------
+    unit_test_cases_dir : str
+        Path string to the location of unit test cases path.
+    target_table : str or None
+        Name of table that we want to run regression for. If None, run all found.
+
+    Returns
+    -------
+    pd.DataFrame
+        A DataFrame that has all the targetted test cases that we need to run.
+    """
+    all_unit_test_cases = sorted(
+        Path(unit_test_cases_dir).rglob("*.{}".format(SUPPORTED_TC_EXT))
+    )
+    logging.info(
+        "## Total number of test cases found: {}".format(len(all_unit_test_cases))
+    )
+
+    # Get test cases df from test cases
+    tc_df = pd.DataFrame({"test_path": all_unit_test_cases})
+    tc_df["table_name"] = tc_df["test_path"].apply(
+        lambda x: x.name.replace(".gds", "")
+    )
+
+    if target_table is not None:
+        tc_df = tc_df[tc_df["table_name"] == target_table]
+
+    if len(tc_df) < 1:
+        logging.error("No test cases remaining after filtering.")
+        exit(1)
+
+    tc_df["run_id"] = range(len(tc_df))
+    return tc_df
+
+
+def aggregate_results(tc_df: pd.DataFrame, results_df: pd.DataFrame, rules_df: pd.DataFrame):
+    """
+    aggregate_results Aggregate the results for all runs.
+
+    Parameters
+    ----------
+    tc_df : pd.DataFrame
+        Dataframe that holds the information about the test cases.
+    results_df : pd.DataFrame
+        Dataframe that holds the information about the unit test rules.
+    rules_df : pd.DataFrame
+        Dataframe that holds the information about all the rules implemented in the rule deck.
+
+    Returns
+    -------
+    pd.DataFrame
+        A DataFrame that has all data analysis aggregated into one.
+    """
+    if len(rules_df) < 1 and len(results_df) < 1:
+        logging.error("## There are no rules for analysis or run.")
+        exit(1)
+    elif len(rules_df) < 1 and len(results_df) > 0:
+        df = results_df
+    elif len(rules_df) > 0 and len(results_df) < 1:
+        df = rules_df
+        for c in ANALYSIS_RULES:
+            df[c] = 0
+    else:
+        df = results_df.merge(rules_df, how="outer", on=["table_name", "rule_name"])
+
+    df[ANALYSIS_RULES] = df[ANALYSIS_RULES].fillna(0)
+    df["in_rule_deck"] = df["in_rule_deck"].fillna(0)
+    df = df.merge(tc_df[["table_name", "run_status"]], how="left", on="table_name")
+
+    df["rule_status"] = "Passed"
```

Comment:
does that mean that we mark rule as passing par default? is there a way to explicit set the status from the actual run instead?

Reply:
We start with pass status, but look the lines below that. It changes each testcase based on the analysis. It's easier and faster to do that in pandas.

Reply:
yep, I understand that, but how do we make sure that the condition after that cover all failure cases (and that we're not leaving something that failed marked as 'pass').

Wouldn't it be possible to mark everything as `unknown` by default, and only mark as `pass` if the test actually passed?

Reply:
@proppy will make that change.

Reply:
done